draft-ietf-httpbis-header-structure.md

```diff
@@ -292,6 +292,12 @@ In HTTP headers, members are separated by a comma with optional whitespace, whil
 Example-DictHeader: en="Applepie", da=*w4ZibGV0w6ZydGU=*
 ~~~
 
+Members whose value is Boolean true omit their value when serialised, unless it has parameters. For example, here both "b" and "c" are true, but "c"'s value is serialised because it has parameters:
```

Comment:
Is the omission a MAY, SHOULD, or MUST?

The decoder logic seems to imply that it is not a MUST, but I think it would be a good idea to clarify what is valid and not.

Reply:
I think it's MAY.

Reply:
It should be a MUST if we want to use this to back port existing headers (e.g., `Cache-Control`).

Reply:
Also, I'm wondering if we can allow parameters on one of these values; e.g., `baz;thing=1`

Reply:
If it's a MUST then you introduce a really weird constraint on the parser, to detect and raise an exception on an otherwise valid input. The finite state machine essentially has parse-bare-item-excluding-true and parse-bare-item-including-true routines.

I think I missed the decision to make back-porting existing headers a goal. Last I heard it was out of scope. Has that switched around?

Reply:
Sorry, I meant MUST encode, not decode.

Reply:
There's a smell. It's a MUST that is syntactically valid, and not caught as an explicit exception by the machinery.

The requirement is on the backported headers themselves – are they allowed to constrain the serialisation beyond the core spec?

Reply:
My two cents go to either:
* encoder MUST omit & decoder MUST enforce
* encoder SHOULD omit

Structured Headers is something new, and I think we can start with a strict rule.

Reply:
I would be more inclined toward "MUST omit" if `x=?0` and absence of `x` were always semantically equivalent.  Is it reasonable to say that Boolean-valued parameters have to default to `?0` in their absence?

Reply:
@martinthomson I'm not sure if I like the idea of switching to a binary logic from the ternary logic that we have now (i.e. true, false or absent).

In case of [priority "response" header](https://kazuho.github.io/draft-kazuho-httpbis-priority/draft-kazuho-httpbis-priority.html#rfc.section.6), absence of `i` parameter is different from `i=?0`. Absence means "refer to what client has suggested," false means "not incremental."

The issue can be solved by having different terms for referring to a "incremental" response and a "non-incremental" response, much like we do with the Cache-Control header field. IIUC, Cache-Control has "pubilc", "private", and absence of both means default.

Though I am not sure if that is a good design pattern, good enough so that it can be enforced across all the users of SH.

Reply:
It's quite common to have stricter requirements on serialisation than on parsers; e.g., see how we handle whitespace in various places. The requirement here is very specifically targeted to allow the back port, just as with whitespace handling.

`x=?0` and absence of `x` are different things; if we want to make them equivalent, that's a different issue, I think (effectively, any value not present is false -- which IMO would be weird). 

I'm about to push an update, PTAL.

Reply:
It still feels kind of strange, but I can implement it. (And probably would have anyway, if that's the "canonical" form.)

Reply:
BTW True with parameters is currently serialised as `x=?1;foo=bar`. Are folks OK with making it consistent so that would be `x;foo=bar`?

Reply:
Ah, I remember why I didn't do that; it would require pretty major surgery on the algorithms. Given that we don't need to back port anything for that, I'm inclined to leave it; never mind.